cody-repo-util.el

```diff
@@ -0,0 +1,182 @@
+    (condition-case err
+        (let* ((expanded (if (string-match "^[a-zA-Z]:" file)
+                             file
+                           (expand-file-name file)))
+               (normalized (replace-regexp-in-string "\\\\" "/" expanded)) ; Normalize Windows paths
+               (uri (if (file-remote-p normalized)
+                        normalized
+                      (concat "file://"
+                              (if (string-match "\\([a-zA-Z]\\):" normalized)
+                                  ;; Handle Windows drive letters
+                                  (concat "/" (downcase (match-string 1 normalized)) "%3A"
+                                          (substring normalized 2))
+                                normalized)))))
+          uri)
+      (error
+       (message "Error converting file to URI: %s" err)
+       (buffer-name buffer)))))
+
+(defun cody--repo-info (project file)
+  "Return repository info for PROJECT and FILE."
+  (let* ((vcs-type (cody--get-vcs-type project file))
+         (relative-path "")
+         (remote-url "")
+         (remote-branch-name "")
+         (repo-root-path (cody--get-repo-root-path project file)))
+    (when repo-root-path
+      (setq relative-path (if (> (length (file-relative-name file repo-root-path)) 0)
+                              (file-relative-name file repo-root-path)
+                            ""))
+      (when (and (eq vcs-type 'perforce) (string-match "/" relative-path))
+        (setq relative-path (substring relative-path (match-end 0))))
+      (setq remote-url (cody--get-remote-repo-url project file))
+      (setq remote-url (cody--url-do-replacements remote-url))
+      (setq remote-branch-name (cody--get-remote-branch-name project file))
+      (when (or (not remote-branch-name)
+                (not (string-match-p remote-branch-name remote-url)))
+        (setq remote-branch-name cody-default-branch-name)))
+    (list :vcs-type vcs-type
+          :remote-url remote-url
+          :remote-branch-name (or remote-branch-name cody--default-branch-name)
+          :relative-path relative-path)))
+
+(defun cody--find-repository-name (project current-file)
+  "Find the repository name for PROJECT and CURRENT-FILE."
+  (when-let ((file-from-repo (or current-file
+                                 (cody--get-root-file-from-first-git-repo project))))
+    (condition-case nil
+        (cody--get-remote-repo-url project file-from-repo)
+      (error
+       (cody--get-simple-repository-name project file-from-repo)))))
+
+(defun cody--get-simple-repository-name (project file)
+  "Get the simple repository name for PROJECT and FILE."
+  (when-let ((repository (vc-call-backend (vc-backend file) 'root file)))
+    (file-name-nondirectory (directory-file-name repository))))
+
+(defun cody--url-do-replacements (remote-url)
+  "Perform URL replacements for REMOTE-URL"
+  (let* ((replacements
+          (split-string cody--remote-url-replacements "\\s*,\\s*"))
```

Comment:
In the cody.el `defcustom` this is `cody-remote-url-replacements` note the  `cody-` versus `cody--`